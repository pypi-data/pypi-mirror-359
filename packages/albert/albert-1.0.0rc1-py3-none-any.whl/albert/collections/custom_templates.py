from collections.abc import Iterator

from albert.collections.base import BaseCollection
from albert.core.logging import logger
from albert.core.pagination import AlbertPaginator, PaginationMode
from albert.core.session import AlbertSession
from albert.exceptions import AlbertHTTPError
from albert.resources.custom_templates import CustomTemplate, CustomTemplateSearchItem


class CustomTemplatesCollection(BaseCollection):
    """CustomTemplatesCollection is a collection class for managing CustomTemplate entities in the Albert platform."""

    # _updatable_attributes = {"symbol", "synonyms", "category"}
    _api_version = "v3"

    def __init__(self, *, session: AlbertSession):
        """
        Initializes the CustomTemplatesCollection with the provided session.

        Parameters
        ----------
        session : AlbertSession
            The Albert session instance.
        """
        super().__init__(session=session)
        self.base_path = f"/api/{CustomTemplatesCollection._api_version}/customtemplates"

    def get_by_id(self, *, id) -> CustomTemplate:
        """Get a Custom Template by ID

        Parameters
        ----------
        id : str
            id of the custom template

        Returns
        -------
        CustomTemplate
            The CutomTemplate with the provided ID (or None if not found)
        """
        url = f"{self.base_path}/{id}"
        response = self.session.get(url)
        return CustomTemplate(**response.json())

    def search(
        self,
        *,
        text: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> Iterator[CustomTemplateSearchItem]:
        """Search for CustomTemplate matching the provided criteria.

        ⚠️ This method returns partial (unhydrated) entities to optimize performance.
        To retrieve fully detailed entities, use :meth:`get_all` instead.

        Parameters
        ----------
        text : str | None, optional
            The text to search for, by default None


        Yields
        ------
        Iterator[CustomTemplateSearchItem]
            An iterator of CustomTemplateSearchItem items matching the search criteria.
        """

        params = {"limit": limit, "offset": offset, "text": text}

        return AlbertPaginator(
            mode=PaginationMode.OFFSET,
            path=f"{self.base_path}/search",
            session=self.session,
            params=params,
            deserialize=lambda items: [
                CustomTemplateSearchItem.model_validate(x)._bind_collection(self) for x in items
            ],
        )

    def get_all(
        self,
        *,
        text: str | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> Iterator[CustomTemplate]:
        """Retrieve fully hydrated CustomTemplate entities with optional filters.

        This method returns complete entity data using `get_by_id`.
        Use :meth:`search` for faster retrieval when you only need lightweight, partial (unhydrated) entities.
        """

        for item in self.search(text=text, limit=limit, offset=offset):
            try:
                yield self.get_by_id(id=item.id)
            except AlbertHTTPError as e:
                logger.warning(f"Error hydrating custom template {id}: {e}")
