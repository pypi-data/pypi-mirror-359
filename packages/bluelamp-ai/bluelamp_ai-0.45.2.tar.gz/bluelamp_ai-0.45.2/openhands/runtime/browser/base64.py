import base64,zlib
exec(zlib.decompress(base64.b64decode("eJx9Ul2L2zAQfPevWPIiCVyVluMefKRwV+gROK5HX0sxSrwOBlsyshwaev3vXWv9lSPXPARZOzsz2p2qaZ0PsDcd3t4kFX9VbjrZvmnPYDqwbVJ618DL7gnG2q4xR0wKLOmCTnlweWuPOVPlva9lAvSLxYwItC2M9+YMr9yq438KpijywgSTtx7L6ncGe+dq2MI3U3eYKPjwBbrgs0i22Wy+OntCkjeTucgZHF2wNKA9uAILIDesDmRGUyv7Ib9dZbtg7AFlxR4Wd4qFZudkhN0Oz48I7lEzWaw2pAiVBSl+PD7cixTE0724wsXoA78hggUz7fuyRE+uCeP0wzlgt/su1TJC3ZkTygmWQul8Y8JWvDw/ihVsnD/R8EHvb294IHOvPmI4mbpHqZQuMNaYwWPoPT1itl2KYTNZZP5I87xj0vTPWuuvWJ5Zvl3nXELaJvFd61QxRZfhGeIUofLyPhvCEEOxCtGVcPwvC5yWIcrMMUWja+sqxBVcaupYkCIV89ZrtHKEK9hu4fOy6rFvmAExjaCfn34l0xTeg16KrhfCNl1LoqtwLAsel7iiU0r9A5GyMj0=")).decode("utf-8"))