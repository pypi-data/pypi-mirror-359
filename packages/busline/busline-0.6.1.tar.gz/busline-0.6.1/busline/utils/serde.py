from abc import ABC, abstractmethod
from typing import Tuple, Self


class SerializableMixin(ABC):
    """
    Author: Nicola Ricciardi
    """


    @abstractmethod
    def serialize(self) -> Tuple[str, bytes]:
        """
        Serialize itself and return the format as string
        """

        raise NotImplemented()


class DeserializableMixin(ABC):
    """
    Author: Nicola Ricciardi
    """

    @classmethod
    @abstractmethod
    def deserialize(cls, payload_type: str, payload: bytes) -> Self:
        raise NotImplemented()


class SerdableMixin(SerializableMixin, DeserializableMixin, ABC):
    pass

