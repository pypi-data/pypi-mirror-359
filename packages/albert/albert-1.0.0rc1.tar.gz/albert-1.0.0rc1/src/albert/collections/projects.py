from collections.abc import Iterator

from albert.collections.base import BaseCollection
from albert.core.logging import logger
from albert.core.pagination import AlbertPaginator, PaginationMode
from albert.core.session import AlbertSession
from albert.exceptions import AlbertHTTPError
from albert.resources.projects import Project, ProjectFilterParams, ProjectSearchItem


class ProjectCollection(BaseCollection):
    """ProjectCollection is a collection class for managing Project entities in the Albert platform."""

    _api_version = "v3"
    _updatable_attributes = {"description", "grid", "metadata", "state"}

    def __init__(self, *, session: AlbertSession):
        """
        Initialize a ProjectCollection object.

        Parameters
        ----------
        session : AlbertSession
            The Albert session instance.
        """
        super().__init__(session=session)
        self.base_path = f"/api/{ProjectCollection._api_version}/projects"

    def create(self, *, project: Project) -> Project:
        """
        Create a new project.

        Parameters
        ----------
        project : Project
            The project to create.

        Returns
        -------
        Optional[Project]
            The created project object if successful, None otherwise.
        """
        response = self.session.post(
            self.base_path, json=project.model_dump(by_alias=True, exclude_unset=True, mode="json")
        )
        return Project(**response.json())

    def get_by_id(self, *, id: str) -> Project:
        """
        Retrieve a project by its ID.

        Parameters
        ----------
        id : str
            The ID of the project to retrieve.

        Returns
        -------
        Project
            The project object if found
        """
        url = f"{self.base_path}/{id}"
        response = self.session.get(url)

        return Project(**response.json())

    def update(self, *, project: Project) -> Project:
        """Update a project.

        Parameters
        ----------
        project : Project
            The updated project object.

        Returns
        -------
        Project
            The updated project object as returned by the server.
        """
        existing_project = self.get_by_id(id=project.id)
        patch_data = self._generate_patch_payload(existing=existing_project, updated=project)
        url = f"{self.base_path}/{project.id}"

        self.session.patch(url, json=patch_data.model_dump(mode="json", by_alias=True))

        return self.get_by_id(id=project.id)

    def delete(self, *, id: str) -> None:
        """
        Delete a project by its ID.

        Parameters
        ----------
        id : str
            The ID of the project to delete.

        Returns
        -------
        None
        """
        url = f"{self.base_path}/{id}"
        self.session.delete(url)

    def search(self, *, params: ProjectFilterParams | None = None) -> Iterator[ProjectSearchItem]:
        """Search for Project matching the provided criteria.

        ⚠️ This method returns partial (unhydrated) entities to optimize performance.
        To retrieve fully detailed entities, use :meth:`get_all` instead.

        Parameters
        ----------
        params : ProjectFilterParams, optional
            Structured query parameters to filter, sort, and paginate projects.

        Returns
        -------
        Iterator[Project]
            An iterator of Project resources.
        """
        params = params or ProjectFilterParams()

        query_params = {
            "limit": params.limit,
            "order": params.order_by.value,
            "text": params.text,
            "sortBy": params.sort_by,
            "status": params.status,
            "marketSegment": params.market_segment,
            "application": params.application,
            "technology": params.technology,
            "createdBy": params.created_by,
            "location": params.location,
            "fromCreatedAt": params.from_created_at,
            "toCreatedAt": params.to_created_at,
            "facetField": params.facet_field,
            "facetText": params.facet_text,
            "containsField": params.contains_field,
            "containsText": params.contains_text,
            "linkedTo": params.linked_to,
            "myProjects": params.my_projects,
            "myRole": params.my_role,
        }

        return AlbertPaginator(
            mode=PaginationMode.OFFSET,
            path=f"{self.base_path}/search",
            session=self.session,
            params=query_params,
            deserialize=lambda items: [
                ProjectSearchItem(**item)._bind_collection(self) for item in items
            ],
        )

    def get_all(self, *, params: ProjectFilterParams | None = None) -> Iterator[Project]:
        """Retrieve fully hydrated Project entities with optional filters.

        This method returns complete entity data using `get_by_id`.
        Use :meth:`search` for faster retrieval when you only need lightweight, partial (unhydrated) entities.

        Parameters
        ----------
        params : ProjectFilterParams, optional
            Structured query parameters to filter, sort, and paginate projects.

        Returns
        -------
        Iterator[Project]
            An iterator of fully hydrated Project entities.
        """
        params = params or ProjectFilterParams()

        for project in self.search(params=params):
            project_id = getattr(project, "albertId", None) or getattr(project, "id", None)
            if not project_id:
                continue

            id = project_id if project_id.startswith("PRO") else f"PRO{project_id}"

            try:
                yield self.get_by_id(id=id)
            except AlbertHTTPError as e:
                logger.warning(f"Error fetching project details {id}: {e}")
