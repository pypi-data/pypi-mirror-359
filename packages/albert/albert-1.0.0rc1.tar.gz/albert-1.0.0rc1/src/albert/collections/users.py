from collections.abc import Iterator

import jwt

from albert.collections.base import BaseCollection
from albert.core.logging import logger
from albert.core.pagination import AlbertPaginator, PaginationMode
from albert.core.session import AlbertSession
from albert.core.shared.enums import Status
from albert.exceptions import AlbertHTTPError
from albert.resources.users import User, UserFilterParams, UserFilterType, UserSearchItem


class UserCollection(BaseCollection):
    """UserCollection is a collection class for managing User entities in the Albert platform."""

    _api_version = "v3"
    _updatable_attributes = {"name", "status", "email", "metadata"}

    def __init__(self, *, session: AlbertSession):
        """
        Initializes the UserCollection with the provided session.

        Parameters
        ----------
        session : AlbertSession
            The Albert session instance.
        """
        super().__init__(session=session)
        self.base_path = f"/api/{UserCollection._api_version}/users"

    def get_current_user(self) -> User:
        """
        Retrieves the current authenticated user.

        Returns
        -------
        User
            The current User object.
        """
        claims = jwt.decode(self.session._access_token, options={"verify_signature": False})
        return self.get_by_id(id=claims["id"])

    def get_by_id(self, *, id: str) -> User:
        """
        Retrieves a User by its ID.

        Parameters
        ----------
        id : str
            The ID of the user to retrieve.

        Returns
        -------
        User
            The User object.
        """
        url = f"{self.base_path}/{id}"
        response = self.session.get(url)
        return User(**response.json())

    def search(self, *, params: UserFilterParams | None = None) -> Iterator[UserSearchItem]:
        """
        Searches for Users matching the provided criteria.

        ⚠️ This method returns partial (unhydrated) search results for performance.
        To retrieve fully detailed objects, use :meth:`get_all` instead.

        Parameters
        ----------
        params : UserFilterParams, optional
            Structured search filters for user listing.

        Returns
        -------
        Iterator[User]
            An iterator of partial User entities.
        """

        params = params or UserFilterParams()
        query_params = params.model_dump(exclude_none=True, by_alias=True)

        return AlbertPaginator(
            mode=PaginationMode.OFFSET,
            path=f"{self.base_path}/search",
            session=self.session,
            params=query_params,
            deserialize=lambda items: [
                UserSearchItem(**item)._bind_collection(self) for item in items
            ],
        )

    def get_all(
        self,
        *,
        limit: int = 100,
        status: Status | None = None,
        type: UserFilterType | None = None,
        id: list[str] | None = None,
        start_key: str | None = None,
    ) -> Iterator[User]:
        """
        Retrieve fully hydrated User entities with optional filters.

        This method uses `get_by_id` to hydrate the results for convenience.
        Use :meth:`search` for better performance.

        Parameters
        ----------
        limit : int
            Max results per page.
        status : Status, optional
            Filter by user status.
        type : UserFilterType
            Attribute name to filter by (e.g., 'role').
        id : list[str], optional
            Values of the attribute to filter on.
        start_key : Optional[str], optional
            The starting point for the next set of results, by default None.

        Returns
        -------
        Iterator[User]
            Fully hydrated User entities.
        """
        params = {
            "limit": limit,
            "status": status,
            "type": type.value if type else None,
            "id": id,
            "startKey": start_key,
        }

        def deserialize(items: list[dict]) -> Iterator[User]:
            for item in items:
                user_id = item.get("albertId")
                if user_id:
                    try:
                        yield self.get_by_id(id=user_id)
                    except AlbertHTTPError as e:
                        logger.warning(f"Error fetching user '{user_id}': {e}")

        return AlbertPaginator(
            mode=PaginationMode.KEY,
            path=self.base_path,
            session=self.session,
            params=params,
            deserialize=deserialize,
        )

    def create(self, *, user: User) -> User:  # pragma: no cover
        """Create a new User

        Parameters
        ----------
        user : User
            The user to create

        Returns
        -------
        User
            The created User
        """

        response = self.session.post(
            self.base_path,
            json=user.model_dump(by_alias=True, exclude_none=True, mode="json"),
        )
        return User(**response.json())

    def update(self, *, user: User) -> User:
        """Update a User entity.

        Parameters
        ----------
        user : User
            The updated User entity.

        Returns
        -------
        User
            The updated User entity as returned by the server.
        """
        # Fetch the current object state from the server or database
        current_object = self.get_by_id(id=user.id)

        # Generate the PATCH payload
        payload = self._generate_patch_payload(existing=current_object, updated=user)

        url = f"{self.base_path}/{user.id}"
        self.session.patch(url, json=payload.model_dump(mode="json", by_alias=True))

        updated_user = self.get_by_id(id=user.id)
        return updated_user
