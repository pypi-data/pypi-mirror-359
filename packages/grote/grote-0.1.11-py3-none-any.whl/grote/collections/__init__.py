from grote.collections.load import LoadComponents
from grote.collections.translate import TranslateComponents

__all__ = [
    "LoadComponents",
    "TranslateComponents",
]
