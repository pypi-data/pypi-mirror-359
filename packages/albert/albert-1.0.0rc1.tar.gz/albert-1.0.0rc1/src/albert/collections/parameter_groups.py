from __future__ import annotations

from collections.abc import Iterator

from albert.collections.base import BaseCollection
from albert.core.logging import logger
from albert.core.pagination import AlbertPaginator, PaginationMode
from albert.core.session import AlbertSession
from albert.core.shared.enums import OrderBy
from albert.exceptions import AlbertHTTPError
from albert.resources.parameter_groups import (
    ParameterGroup,
    ParameterGroupSearchItem,
    PGType,
)
from albert.utils._patch import generate_parameter_group_patches


class ParameterGroupCollection(BaseCollection):
    """ParameterGroupCollection is a collection class for managing ParameterGroup entities in the Albert platform."""

    _api_version = "v3"
    _updatable_attributes = {"name", "description", "metadata"}
    # To do: Add the rest of the allowed attributes

    def __init__(self, *, session: AlbertSession):
        """A collection for interacting with Albert parameter groups.

        Parameters
        ----------
        session : AlbertSession
            The Albert session to use for making requests.
        """
        super().__init__(session=session)
        self.base_path = f"/api/{ParameterGroupCollection._api_version}/parametergroups"

    def get_by_id(self, *, id: str) -> ParameterGroup:
        """Get a parameter group by its ID.

        Parameters
        ----------
        id : str
            The ID of the parameter group to retrieve.

        Returns
        -------
        ParameterGroup
            The parameter group with the given ID.
        """
        path = f"{self.base_path}/{id}"
        response = self.session.get(path)
        return ParameterGroup(**response.json())

    def get_by_ids(self, *, ids: list[str]) -> ParameterGroup:
        url = f"{self.base_path}/ids"
        batches = [ids[i : i + 100] for i in range(0, len(ids), 100)]
        return [
            ParameterGroup(**item)
            for batch in batches
            for item in self.session.get(url, params={"id": batch}).json()["Items"]
        ]

    def search(
        self,
        *,
        text: str | None = None,
        types: PGType | list[PGType] | None = None,
        order_by: OrderBy = OrderBy.DESCENDING,
        limit: int = 25,
        offset: int | None = None,
    ) -> Iterator[ParameterGroupSearchItem]:
        """Search for Parameter Groups matching the given criteria.

        Parameters
        ----------
        text : str | None, optional
            Text to search for, by default None
        types : PGType | list[PGType] | None, optional
            Filer the returned Parameter Groups by Type, by default None
        order_by : OrderBy, optional
            The order in which to return results, by default OrderBy.DESCENDING

        Yields
        ------
        Iterator[ParameterGroup]
            An iterator of Parameter Groups matching the given criteria.
        """

        params = {
            "limit": limit,
            "offset": offset,
            "order": order_by.value,
            "text": text,
            "types": [types] if isinstance(types, PGType) else types,
        }

        return AlbertPaginator(
            mode=PaginationMode.OFFSET,
            path=f"{self.base_path}/search",
            session=self.session,
            params=params,
            deserialize=lambda items: [
                ParameterGroupSearchItem(**item)._bind_collection(self) for item in items
            ],
        )

    def get_all(
        self,
        *,
        text: str | None = None,
        types: PGType | list[PGType] | None = None,
        order_by: OrderBy = OrderBy.DESCENDING,
        limit: int = 25,
        offset: int | None = None,
    ) -> Iterator[ParameterGroup]:
        """Search and hydrate all Parameter Groups matching the given criteria.

        Parameters
        ----------
        text : str | None, optional
            Text to search for, by default None.
        types : PGType | list[PGType] | None, optional
            Filter the returned Parameter Groups by Type, by default None.
        order_by : OrderBy, optional
            The order in which to return results, by default OrderBy.DESCENDING.
        limit : int, optional
            Page size for each search request, by default 25.
        offset : int | None, optional
            Offset to start from, by default None.

        Yields
        ------
        Iterator[ParameterGroup]
            An iterator of fully hydrated Parameter Groups.
        """
        for item in self.search(
            text=text,
            types=types,
            order_by=order_by,
            limit=limit,
            offset=offset,
        ):
            try:
                # Currently, the API is not returning Metadata, Tags, Documents, and ACL for the get_by_ids endpoint, so we need to fetch individually until that is fixed
                yield self.get_by_id(id=item.id)
            except AlbertHTTPError as e:  # pragma: no cover
                logger.warning(f"Error fetching parameter group {item.id}: {e}")

    def delete(self, *, id: str) -> None:
        """Delete a parameter group by its ID.

        Parameters
        ----------
        id : str
            The ID of the parameter group to delete
        """
        path = f"{self.base_path}/{id}"
        self.session.delete(path)

    def create(self, *, parameter_group: ParameterGroup) -> ParameterGroup:
        """Create a new parameter group.

        Parameters
        ----------
        parameter_group : ParameterGroup
            The parameter group to create.

        Returns
        -------
        ParameterGroup
            The created parameter group.
        """

        response = self.session.post(
            self.base_path,
            json=parameter_group.model_dump(by_alias=True, exclude_none=True, mode="json"),
        )
        return ParameterGroup(**response.json())

    def get_by_name(self, *, name: str) -> ParameterGroup | None:
        """Get a parameter group by its name.

        Parameters
        ----------
        name : str
            The name of the parameter group to retrieve.

        Returns
        -------
        ParameterGroup | None
            The parameter group with the given name, or None if not found.
        """
        matches = self.get_all(text=name)
        # TODO: optimize with explicit hydrate() after self.search()
        for m in matches:
            if m.name.lower() == name.lower():
                return m
        return None

    def update(self, *, parameter_group: ParameterGroup) -> ParameterGroup:
        """Update a parameter group.

        Parameters
        ----------
        parameter_group : ParameterGroup
            The updated ParameterGroup. The ParameterGroup must have an ID.

        Returns
        -------
        ParameterGroup
            The updated ParameterGroup as returned by the server.
        """
        existing = self.get_by_id(id=parameter_group.id)
        path = f"{self.base_path}/{existing.id}"

        base_payload = self._generate_patch_payload(
            existing=existing, updated=parameter_group, generate_metadata_diff=True
        )

        general_patches, new_parameter_values, enum_patches = generate_parameter_group_patches(
            initial_patches=base_payload,
            updated_parameter_group=parameter_group,
            existing_parameter_group=existing,
        )

        # add new parameters
        new_param_url = f"{self.base_path}/{parameter_group.id}/parameters"
        if len(new_parameter_values) > 0:
            self.session.put(
                url=new_param_url,
                json=[
                    x.model_dump(mode="json", by_alias=True, exclude_none=True)
                    for x in new_parameter_values
                ],
            )
        new_param_sequences = [x.sequence for x in new_parameter_values]
        # handle enum updates
        for sequence, ep in enum_patches.items():
            if sequence in new_param_sequences:
                # we don't need to handle enum updates for new parameters
                continue
            if len(ep) > 0:
                enum_url = f"{self.base_path}/{parameter_group.id}/parameters/{sequence}/enums"
                self.session.put(
                    url=enum_url,
                    json=ep,
                )
        if len(general_patches.data) > 0:
            # patch the general patches
            self.session.patch(
                url=path,
                json=general_patches.model_dump(mode="json", by_alias=True, exclude_none=True),
            )

        return self.get_by_id(id=parameter_group.id)
