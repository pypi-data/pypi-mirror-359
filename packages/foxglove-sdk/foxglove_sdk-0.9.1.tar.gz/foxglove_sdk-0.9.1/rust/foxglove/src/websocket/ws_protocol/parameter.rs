//! Parameter types.

use std::collections::BTreeMap;

use base64::prelude::*;
use serde::{Deserialize, Serialize};
use serde_with::serde_as;

/// Error encountered while trying to decide a base64-encoded byte array parameter value.
#[derive(Debug, thiserror::Error)]
pub enum DecodeError {
    /// Parameter is not a byte-array.
    #[error("Parameter is not a byte array")]
    WrongType,
    /// Invalid base64.
    #[error(transparent)]
    Base64(#[from] base64::DecodeError),
}

/// A parameter type.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ParameterType {
    /// A byte array, encoded as a base64-encoded string.
    ByteArray,
    /// A decimal or integer value that can be represented as a `float64`.
    Float64,
    /// An array of decimal or integer values that can be represented as `float64`s.
    Float64Array,
}

/// A parameter value.
#[serde_as]
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(untagged)]
pub enum ParameterValue {
    /// A decimal or integer value.
    Number(f64),
    /// A boolean value.
    Bool(bool),
    /// A string value.
    ///
    /// For parameters of type [`ParameterType::ByteArray`], this is a base64 encoding of the byte
    /// array.
    String(String),
    /// An array of parameter values.
    Array(Vec<ParameterValue>),
    /// An associative map of parameter values.
    Dict(BTreeMap<String, ParameterValue>),
}

/// Informs the client about a parameter.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Parameter {
    /// The name of the parameter.
    pub name: String,
    /// The parameter type.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub r#type: Option<ParameterType>,
    /// The parameter value.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub value: Option<ParameterValue>,
}

impl Parameter {
    /// Creates a new parameter with no value or type.
    pub fn empty(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            r#type: None,
            value: None,
        }
    }

    /// Creates a new parameter with a float64 value.
    pub fn float64(name: impl Into<String>, value: f64) -> Self {
        Self {
            name: name.into(),
            r#type: Some(ParameterType::Float64),
            value: Some(ParameterValue::Number(value)),
        }
    }

    /// Creates a new parameter with a float64 array value.
    pub fn float64_array(name: impl Into<String>, values: impl IntoIterator<Item = f64>) -> Self {
        Self {
            name: name.into(),
            r#type: Some(ParameterType::Float64Array),
            value: Some(ParameterValue::Array(
                values.into_iter().map(ParameterValue::Number).collect(),
            )),
        }
    }

    /// Creates a new parameter with a string value.
    pub fn string(name: impl Into<String>, value: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            r#type: None,
            value: Some(ParameterValue::String(value.into())),
        }
    }

    /// Creates a new parameter with a byte array value.
    pub fn byte_array(name: impl Into<String>, data: &[u8]) -> Self {
        let value = BASE64_STANDARD.encode(data);
        Self {
            name: name.into(),
            r#type: Some(ParameterType::ByteArray),
            value: Some(ParameterValue::String(value)),
        }
    }

    /// Creates a new parameter with a boolean value.
    pub fn bool(name: impl Into<String>, value: bool) -> Self {
        Self {
            name: name.into(),
            r#type: None,
            value: Some(ParameterValue::Bool(value)),
        }
    }

    /// Creates a new parameter with a dictionary value.
    pub fn dict(name: impl Into<String>, value: BTreeMap<String, ParameterValue>) -> Self {
        Self {
            name: name.into(),
            r#type: None,
            value: Some(ParameterValue::Dict(value)),
        }
    }

    /// Decodes a byte array parameter.
    ///
    /// Returns None if the parameter is unset/empty. Returns an error if the parameter value is
    /// not a byte array, or if it is not a valid base64 encoding.
    pub fn decode_byte_array(&self) -> Result<Option<Vec<u8>>, DecodeError> {
        match (self.r#type, self.value.as_ref()) {
            (Some(ParameterType::ByteArray), Some(ParameterValue::String(s))) => {
                Some(BASE64_STANDARD.decode(s).map_err(DecodeError::Base64)).transpose()
            }
            (_, None) => Ok(None),
            _ => Err(DecodeError::WrongType),
        }
    }
}

#[cfg(test)]
mod tests {
    use assert_matches::assert_matches;

    use super::*;

    #[test]
    fn test_empty() {
        insta::assert_json_snapshot!(Parameter::empty("test"));
    }

    #[test]
    fn test_float() {
        insta::assert_json_snapshot!(Parameter::float64("f64", 1.23));
    }

    #[test]
    fn test_float_array() {
        insta::assert_json_snapshot!(Parameter::float64_array("f64[]", [1.23, 4.56]));
    }

    #[test]
    fn test_string() {
        insta::assert_json_snapshot!(Parameter::string("string", "howdy"));
    }

    #[test]
    fn test_byte_array() {
        insta::assert_json_snapshot!(Parameter::byte_array("byte[]", &[0x10, 0x20, 0x30]));
    }

    #[test]
    fn test_decode_byte_array() {
        let param = Parameter::byte_array("bytes", b"123");
        let decoded = param.decode_byte_array().unwrap().unwrap();
        assert_eq!(decoded, b"123".to_vec());

        // invalid base64 value
        let param = Parameter {
            name: "invalid".into(),
            r#type: Some(ParameterType::ByteArray),
            value: Some(ParameterValue::String("!!".into())),
        };
        let result = param.decode_byte_array();
        assert_matches!(result, Err(DecodeError::Base64(_)));

        // it's a string, not a byte array
        let param = Parameter::string("string", "eHl6enk=");
        let result = param.decode_byte_array();
        assert_matches!(result, Err(DecodeError::WrongType));

        // unset
        let param = Parameter {
            name: "unset".into(),
            r#type: Some(ParameterType::ByteArray),
            value: None,
        };
        let result = param.decode_byte_array();
        assert_matches!(result, Ok(None));

        // unset of a different type
        let param = Parameter {
            name: "unset".into(),
            r#type: None,
            value: None,
        };
        let result = param.decode_byte_array();
        assert_matches!(result, Ok(None));
    }

    #[test]
    fn test_bool() {
        insta::assert_json_snapshot!(Parameter::bool("bool", true));
    }

    #[test]
    fn test_dict() {
        insta::assert_json_snapshot!(Parameter::dict(
            "outer",
            maplit::btreemap! {
                "bool".into() => ParameterValue::Bool(false),
                "number".into() => ParameterValue::Number(1.23),
                "nested".into() => ParameterValue::Dict(
                    maplit::btreemap! {
                        "inner".into() => ParameterValue::Number(1.0),
                    }
                )
            }
        ));
    }
}
