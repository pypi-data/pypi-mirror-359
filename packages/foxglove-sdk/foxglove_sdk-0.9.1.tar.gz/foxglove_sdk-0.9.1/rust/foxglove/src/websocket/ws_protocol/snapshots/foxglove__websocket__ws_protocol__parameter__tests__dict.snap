---
source: rust/foxglove/src/websocket/ws_protocol/parameter.rs
expression: "Parameter::dict(\"outer\", maplit::btreemap!\n{\n    \"bool\".into() => ParameterValue::Bool(false), \"number\".into() =>\n    ParameterValue::Number(1.23), \"nested\".into() =>\n    ParameterValue::Dict(maplit::btreemap!\n    { \"inner\".into() => ParameterValue::Number(1.0), })\n})"
---
{
  "name": "outer",
  "value": {
    "bool": false,
    "nested": {
      "inner": 1.0
    },
    "number": 1.23
  }
}
